sasquatch: 4.4 -> 4.5.1-1

vlaci opened the pull request:
Required for #217836

###### Description of changes

Changed using ONEKEY [fork](https://github.com/onekey-sec/sasquatch) of sasquatch which is up-to-date with squashfs-tools 4.5.1 and the Darwin compatibility patch shipped in Nixpkgs applies cleanly as well.

I have chosen to override the `squashfsTools` derivation, as this fork is entirely compatible with that version. I am open to critique on this. I wasn't sure copying from squashfstools or lifting a `generic.nix` base derivation out of it is the better solution. Existing NixOS test is also nice to have as a validation, although it may be overkill here.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [x] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

vlaci committed:
sasquatch: 4.4 -> 4.5.1-4

Changed using ONEKEY fork of sasquatch which is up-to-date with
squashfs-tools 4.5.1 and the Darwin compatibility patch used to be
shipped in Nixpkgs applies cleanly as well.

The updated version also provides a big-endian compatible build as
well to support even more insane squashfs dialects.

vlaci commented:
Ah, figured keeping the existing test doesn't make sense as it will just use squashfs-tools anyway.

AkechiShiro commented:
Any news, any blockage ? Or any review from @Pamplemousse if he is still the right contact for this PR ?

vlaci commented:
Some questions got answerred: I shouldn't depend on squashFsTools at all, as it got updated to 4.6.1, and it is unlikely I want to rebase sasquatch the near future.

vlaci commented:
As `squashfsTools` derivation is now incompatible with previous versions, I have recreated the derivation from scratch.

Pamplemousse commented:
@AkechiShiro Sorry I missed your message earlier.

I have stopped using this package in years... I am not well placed to test.
Also, could you remove me from the maintainers please?

vlaci committed:
sasquatch: removing pamplemousse from maintainers

Asked for their removal in
https://github.com/NixOS/nixpkgs/pull/219169#issuecomment-1597379517

vlaci commented:
> @AkechiShiro Sorry I missed your message earlier.
> 
> I have stopped using this package in years... I am not well placed to test. Also, could you remove me from the maintainers please?

Done

qkaiser commented:
This PR has been stuck for the last 6 months waiting for a review from someone who explicitly requested to be removed from the maintainers list.

Who should review this then ? How can we make it move forward ?

Asking cause it's the last item missing in out meta-issue to get unblob in nixpkgs  https://github.com/NixOS/nixpkgs/issues/217836

AkechiShiro commented:
This PR needs a rebase, first of all on master @vlaci could you please rebase please ?

vlaci commented:
Happy new year everyone :)
Rebased on master.

AkechiShiro commented:
Thanks @vlaci, there seems to be a typo in the meta attribute of the package, please fix it, so that the ofborg evaluation test passes, see more details : https://gist.github.com/GrahamcOfBorg/6d11645476eaaef9b7fc87c573d4a188, `meta.mainprogram` should be `meta.mainProgram`. 

I cannot help further as I'm inexperienced in reviews @qkaiser, maybe mentioning the pull request as ready for review here, should bring this PR to more experienced reviewers : https://discourse.nixos.org/t/prs-ready-for-review/3032 and hopefully help.

qkaiser commented:
@AkechiShiro thanks for the heads up, I sent a reply to the Discourse thread asking for a review on this PR

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/1016/87

tnias reviewed (approved):
I ran nixpkgs-review and sucessfully used both x86_64 and aarch64 to extract an example squashfs I found somewhere on my disk.
For me the diffoscope tests are broken on master on x86_64 even without this patch.

---

Result of `nixpkgs-review pr 219169` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>3 packages failed to build:</summary>
  <ul>
    <li>diffoscope</li>
    <li>diffoscope.dist</li>
    <li>diffoscope.man</li>
  </ul>
</details>
<details>
  <summary>10 packages built:</summary>
  <ul>
    <li>binwalk (python311Packages.binwalk)</li>
    <li>binwalk.dist (python311Packages.binwalk.dist)</li>
    <li>python310Packages.binwalk</li>
    <li>python310Packages.binwalk-full</li>
    <li>python310Packages.binwalk-full.dist</li>
    <li>python310Packages.binwalk.dist</li>
    <li>python311Packages.binwalk-full</li>
    <li>python311Packages.binwalk-full.dist</li>
    <li>sasquatch</li>
    <li>sasquatch-v4be</li>
  </ul>
</details>

Result of `nixpkgs-review pr 219169` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>13 packages built:</summary>
  <ul>
    <li>binwalk (python311Packages.binwalk)</li>
    <li>binwalk.dist (python311Packages.binwalk.dist)</li>
    <li>diffoscope</li>
    <li>diffoscope.dist</li>
    <li>diffoscope.man</li>
    <li>python310Packages.binwalk</li>
    <li>python310Packages.binwalk-full</li>
    <li>python310Packages.binwalk-full.dist</li>
    <li>python310Packages.binwalk.dist</li>
    <li>python311Packages.binwalk-full</li>
    <li>python311Packages.binwalk-full.dist</li>
    <li>sasquatch</li>
    <li>sasquatch-v4be</li>
  </ul>
</details>

Mic92 commented:
What does the bigendian variant do? Do we need both?

tilpner commented on pkgs/tools/filesystems/sasquatch/default.nix:
```diff
@@ -1,76 +1,65 @@
-  pname = "sasquatch";
-  version = "4.4";
+  drv = stdenv.mkDerivation rec {
+    pname = "sasquatch${lib.optionalString bigEndian "-v4be"}";
```
This repeats the version number (v4), that's already appended to pname, so the full name would be `sasquatch-v4be-4.5.1.-4`. Perhaps it should just be `sasquatch-be`?

qkaiser replied:
The `v4be` indicates the squashfs [filesystem version](https://docs.kernel.org/filesystems/squashfs.html) while the `4.5.1-4` indicates the squashfs-tools version sasquatch is based on. These versions are unrelated.

qkaiser commented:
> What does the bigendian variant do? Do we need both?

sasquatch is built on top of squashfs-tools to support non standard squashfs filesystems. By the standard, squashfs version 4.0 filesystems are fixed little-endian. There are no big-endian 4.0 filesystems. However, embedded vendors love to mess with standards. So instead of adapting squashfs-tools code, we build a version with `__BYTE_ORDER` swapped which is the `sasquatch-v4be` binary. This binary can handle non-standard squashfs v4 big-endian filesystems.

More information about the big-endian variant can be found at https://github.com/onekey-sec/sasquatch/commit/0c65f00de7c7cc3382302256fddd06ba10ed7b01

Here's a small preview of what vendors do with squashfs. Cells with a star (*) are standard.


type | v2​ | v3​ | v4 | DDWRT v3​ | Broadcom v3​ | ??? v3​ | AVM​ | Netgear​ 
-------|-----|-----|-----|-----------------|---------------------|----------|--------|------------|
magic (LE) | hsqs*​ | hsqs*​ | hsqs*​ | hsqt​ | shsq​  | zlqs​ | Header in BE, but chunks in LE.​ | Let‘s use a non standard LZMA + XZ compression !​ |
magic (BE)​ | sqsh*​ | sqsh*​ | sqsh​ | tqsh​ | qshs​ | sqlz​ | / | / |

AkechiShiro commented:
Hello @smancill I think there was a review that approved the package, I guess running `nixpkgs-review` on this PR and testing the tool and leaving a comment stating how it went could help move the PR forward. 

I don't know however if a rebase is needed to make sure that the current package still build with the latest upstream in Nixpkgs.

AkechiShiro commented:
@vlaci there might be a dead link as a source for the current version : 
![image](https://github.com/user-attachments/assets/be4340de-ceed-415a-a740-199e7b27d2fc)

I cannot find the repo on GitHub, maybe there is new unblob that is being worked on that doesn't depend on `sasquatch` ?

vlaci commented:
> @vlaci there might be a dead link as a source for the current version : ![image](https://private-user-images.githubusercontent.com/14914796/363509699-be4340de-ceed-415a-a740-199e7b27d2fc.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjU4Njc5NzIsIm5iZiI6MTcyNTg2NzY3MiwicGF0aCI6Ii8xNDkxNDc5Ni8zNjM1MDk2OTktYmU0MzQwZGUtY2VlZC00MTVhLWE3NDAtMTk5ZTdiMjdkMmZjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MDklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTA5VDA3NDExMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTE2Njc1ZTdkNmJiYTg0NTFiNGVkZTRhOGUyMzdlOWExODZkZDIyMzg0ZjU5ZTRiNDU0OGQ4MzE1MDliZmExMGEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.MGZsBAnqFjQmmt3cwFcqSMtUt9NsGzX0Aq_D5kR2P7E)
> 
> I cannot find the repo on GitHub, maybe there is new unblob that is being worked on that doesn't depend on `sasquatch` ?
> 
> I only found this repo : https://github.com/onekey-sec/sasquatch
> 
> EDIT :
> 
> * `nixpkgs-review` result :
> 
> ```
> 1 package failed to build:
> sasquatch-v4be
> 
> 14 packages built:
> binwalk binwalk.dist diffoscope diffoscope.dist diffoscope.man python310Packages.binwalk python310Packages.binwalk-full python310Packages.binwalk-full.dist python310Packages.binwalk.dist python311Packages.binwalk python311Packages.binwalk-full python311Packages.binwalk-full.dist python311Packages.binwalk.dist sasquatch
> 
> error: build log of '/nix/store/40nfx0yjl4q7965p05dkqgavd9xjbh7j-sasquatch-v4be-4.5.1-4.drv^*' is not available
> error: build log of '/nix/store/bw99kzndaqc2r8bykzcjc34z91a964j0-sasquatch-v4be-4.5.1-4' is not available
> ```

It must have been a github api error, as the source is there...

vlaci commented:
IDK if I should move these from `all-packages` accorving to nixpkgs-vet.

vlaci committed:
sasquatch: move to pkgs/by-name

vlaci committed:
sasquatch: format package.nix to match RFC style

AkechiShiro commented:
I believe there is a typo in the package @vlaci 

```nix
       error: evaluation aborted with the following error message: 'Failed to evaluate sasquatch-4.5.1-4: «unknown-meta»: has an invalid meta attrset:
         - key 'meta.mainprogram' is unrecognized; expected one of:
         ['available', 'badPlatforms', 'branch', 'broken', 'changelog', 'description', 'downloadPage', 'executables', 'homepage', 'hydraPlatforms', 'insecure', 'isBuildPythonPackage', 'isFcitxEngine', 'isGutenprint', 'isHydraChannel', 'isIbusEngine', 'knownVulnerabilities', 'license', 'longDescription', 'mainProgram', 'maintainers', 'maxSilent', 'name', 'outputsToInstall', 'pkgConfigModules', 'platforms', 'position', 'priority', 'schedulingPriority', 'sourceProvenance', 'tag', 'tests', 'timeout', 'unfree', 'unsupported', 'version']
       '
```

`mainprogram` -> `mainProgram`, if you can also solve the conflict and rebase the PR please ?

vlaci commented:
> I believe there is a typo in the package @vlaci
> 
> ```nix
>        error: evaluation aborted with the following error message: 'Failed to evaluate sasquatch-4.5.1-4: «unknown-meta»: has an invalid meta attrset:
>          - key 'meta.mainprogram' is unrecognized; expected one of:
>          ['available', 'badPlatforms', 'branch', 'broken', 'changelog', 'description', 'downloadPage', 'executables', 'homepage', 'hydraPlatforms', 'insecure', 'isBuildPythonPackage', 'isFcitxEngine', 'isGutenprint', 'isHydraChannel', 'isIbusEngine', 'knownVulnerabilities', 'license', 'longDescription', 'mainProgram', 'maintainers', 'maxSilent', 'name', 'outputsToInstall', 'pkgConfigModules', 'platforms', 'position', 'priority', 'schedulingPriority', 'sourceProvenance', 'tag', 'tests', 'timeout', 'unfree', 'unsupported', 'version']
>        '
> ```
> 
> `mainprogram` -> `mainProgram`, if you can also solve the conflict and rebase the PR please ?

Thanks, this error did not emerge on my Nix version.
Also, I botched up a previous patch, hence the conflict. Adjusted.

AkechiShiro commented on pkgs/top-level/all-packages.nix:
```diff
@@ -12301,7 +12301,7 @@ with pkgs;
   sanctity = callPackage ../tools/misc/sanctity { };
 
-  sasquatch = callPackage ../tools/filesystems/sasquatch { };
+  sasquatch-v4be = callPackage ../by-name/sa/sasquatch/package.nix { bigEndian = true; };
```
@vlaci could you please maybe move this into `by-name/sa/sasquatch-v4be` as well ?

vlaci replied:
Is this how you'd like it?
https://github.com/NixOS/nixpkgs/pull/219169/files#diff-276d44079aa310eda1d4e6bc92d2ffedcf4c0cf9fb5c5f988502809efc60089fR3

AkechiShiro replied:
Yep thanks a lot for your work on this !

AkechiShiro commented:
Tested a build (`nixpkgs-review` seems a bit confused however) with the new changes, seems to be working alright.

Output of `nixpkgs-review` : 
```
8 packages removed:
binwalk (†2.4.2) diffoscope (†277) ghidraninja-ghidra-scripts-unstable (†2020-10-07) python3.11-binwalk (†2.4.2) python3.11-binwalk-full (†2.4.2) python3.12-binwalk (†2.4.2) python3.12-binwalk-full (†2.4.2) sasquatch (†4.4)

Nothing to be built.
```
I ran the build manually.

AkechiShiro reviewed (changes requested):
Just waiting for `sasquatch-v4be` to be moved from `top-level/all-packages.nix` into `by-name/sa/sasquatch-v4be`, then I will approve this PR on my side, I can't really test however on nix `darwin`

vlaci commented:
> Just waiting for `sasquatch-v4be` to be moved from `top-level/all-packages.nix` into `by-name/sa/sasquatch-v4be`, then I will approve this PR on my side, I can't really test however on nix `darwin`

I've created a draft PR for unblob pointing to this branch, so its tests will run for darwin as well: https://github.com/onekey-sec/unblob/pull/946/checks

AkechiShiro commented:
I don't think the `ghidra` failure is related : 
```
1 package failed to build:
ghidra-extensions.ghidraninja-ghidra-scripts

13 packages built:
binwalk binwalk.dist diffoscope diffoscope.dist diffoscope.man python311Packages.binwalk python311Packages.binwalk-full python311Packages.binwalk-full.dist python311Packages.binwalk.dist python312Packages.binwalk-full python312Packages.binwalk-full.dist sasquatch sasquatch-v4be
```

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/4576

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/1969

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/1976

D3vil0p3r commented on pkgs/by-name/sa/sasquatch/package.nix:
```diff
@@ -0,0 +1,72 @@
+}:
+
+let
+  drv = stdenv.mkDerivation rec {
```
```suggestion
  drv = stdenv.mkDerivation (finalAttrs: {
```

D3vil0p3r commented on pkgs/by-name/sa/sasquatch/package.nix:
```diff
@@ -0,0 +1,72 @@
+
+    src = fetchFromGitHub {
+      owner = "onekey-sec";
+      repo = pname;
+      rev = "${pname}-v${version}";
```
```suggestion
      repo = "sasquatch";
      rev = "sasquatch-v${finalAttrs.version}";
```

vlaci replied:
Nice idea to use fixed-point attribute here, but then shouldn't `hash` also come from an input argument? :thinking:

D3vil0p3r replied:
No. My proposals should make the build working correctly.

vlaci replied:
I understand, but IMO `overrideAttrs (_: { version = "foo"; })` will lead to surprising results, as it will affect src fetching

vlaci replied:
Am I overthinking?

D3vil0p3r replied:
Uhm dunno... feel free to not merge the commit suggestion.

vlaci replied:
Checked that there are many examples in nixpkgs already, so it is fine by me

D3vil0p3r commented on pkgs/by-name/sa/sasquatch/package.nix:
```diff
@@ -0,0 +1,72 @@
+      mv $out/bin/sasquatch{,-v4be}
+    '';
+
+    meta = with lib; {
```
```suggestion
    meta = {
```

D3vil0p3r commented on pkgs/by-name/sa/sasquatch/package.nix:
```diff
@@ -0,0 +1,72 @@
+    meta = with lib; {
+      homepage = "https://github.com/onekey-sec/sasquatch";
+      description = "Set of patches to the standard unsquashfs utility (part of squashfs-tools) that attempts to add support for as many hacked-up vendor-specific SquashFS implementations as possible";
+      license = licenses.gpl2Plus;
+      maintainers = with maintainers; [ vlaci ];
+      platforms = platforms.unix;
```
```suggestion
      license = lib.licenses.gpl2Plus;
      maintainers = with lib.maintainers; [ vlaci ];
      platforms = lib.platforms.unix;
```

D3vil0p3r commented on pkgs/by-name/sa/sasquatch/package.nix:
```diff
@@ -0,0 +1,72 @@
+      platforms = platforms.unix;
+      mainProgram = if bigEndian then "sasquatch-v4be" else "sasquatch";
+    };
+  };
```
```suggestion
  });
```

D3vil0p3r reviewed (changes requested):
My 2 cents

D3vil0p3r commented:
@vlaci why do we get those two ofborg-eval errors in check list below?

vlaci commented:
> @vlaci why do we get those two ofborg-eval errors in check list below?

Silly me, reintroduced a previously fixed bug during last rebase :(

smancill commented:
Result of `nixpkgs-review pr 219169` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built:</summary>
  <ul>
    <li>sasquatch</li>
    <li>sasquatch-v4be</li>
  </ul>
</details>
